code/framework/src/integrations/client/instance.cpp

```diff
@@ -268,4 +322,75 @@ namespace Framework::Integrations::Client {
 
         Logging::GetLogger(FRAMEWORK_INNER_CLIENT)->debug("Game sync networking messages registered");
     }
+
+    void Instance::DownloadsAssetsFromConnectedServer() {
+        const auto net = GetNetworkingEngine()->GetNetworkClient();
+
+        // Make sure we're connected to the server already, otherwise bail with warning
+        if (net->GetConnectionState() != Framework::Networking::CONNECTED) {
+            Logging::GetLogger(FRAMEWORK_INNER_CLIENT)->warn("We can't download assets if we are not connected to the server yet!");
+            return;
+        }
+
+        Logging::GetLogger(FRAMEWORK_INNER_CLIENT)->debug("Setting up asset downloads...");
+        const auto serverHash = Framework::Utils::Hashing::CalculateCRC32(_currentState._host + ":" + std::to_string(_currentState._port));
+        const auto cacheDir   = fmt::format("cache\\{}", serverHash);
+        const auto streamer   = net->GetAssetStreamer();
+        SetAssetCachePath(cacheDir);
+        streamer->SetApplicationDirectory(cacheDir.c_str());
+        auto folderHandle = cppfs::fs::open(cacheDir);
+
+        // Ensure we stop existing downloads since the server has pushed new changes already
+        if (_downloadStatus.downloading) {
+            net->GetFileListTransfer()->CancelReceive(_downloadStatus.setID);
+            _downloadStatus = {};
+        }
+
+        if (!folderHandle.exists()) {
+            if (folderHandle.createDirectory()) {
+                Logging::GetLogger(FRAMEWORK_INNER_CLIENT)->debug("Client asset cache: {}", serverHash);
+            }
+            else {
+                Logging::GetLogger(FRAMEWORK_INNER_CLIENT)->warn("Could not create folder for client asset cache: {}", cacheDir);
+                Logging::GetLogger(FRAMEWORK_INNER_CLIENT)->warn("Skip downloading assets.");
+                Logging::GetLogger(FRAMEWORK_INNER_CLIENT)->flush();
+
+                // Ensure we finish the download flow gracefully
+                OnAssetsDownloaded(false);
+            }
+        }
+        Logging::GetLogger(FRAMEWORK_INNER_CLIENT)->flush();
+
+        _downloadStatus.setID = streamer->DownloadFromSubdirectory(nullptr, nullptr, true, net->GetPeer()->GetSystemAddressFromIndex(0), new AssetDownloadFileProgress(this), HIGH_PRIORITY, 2, nullptr);
```

Comment:
Isn't there an issue here where we would still try to call `DownloadFromSubdirectory` when we failed creating the directory?